libkineto/src/CuptiActivityProfiler.cpp

```diff
@@ -1185,6 +1196,24 @@ const time_point<system_clock> CuptiActivityProfiler::performRunLoopStep(
         LOG(INFO) << "Tracing complete.";
         VLOG_IF(1, currentIter > 0) << "This state change was invoked by application's step() call";
 
+        if (currentIter > 0) {
+          if (!stopByIterThread) {
+            std::lock_guard<std::mutex> guard(mutex_);
+            if (!stopByIterThread) {
+              stopByIterThread = new std::thread([collection_done, this, now](){
+                if (libkineto::api().client()) {
```

Comment:
having this code duplicated may be a problem. How about we add a lambda to cleanup the state.

```
auto stopTraceCallback = [collection_done, this, now]](){
        if (libkineto::api().client()) {
          libkineto::api().client()->stop();
        }

#if defined(HAS_CUPTI) || defined(HAS_ROCTRACER)
        ecs_.cupti_stopped_early = cupti_.stopCollection;
#endif // HAS_CUPTI || HAS_ROCTRACER

        std::lock_guard<std::mutex> guard(mutex_);
        stopTraceInternal(now);
        VLOG_IF(0, collection_done) << "Reached profile end time";
        UST_LOGGER_MARK_COMPLETED(kCollectionStage);
}
```

Then you either directly call the function or dispatch it to the thread